src/cjs_parse.rs

```diff
@@ -0,0 +1,1239 @@
+  }
+}
+
+fn get_expr_ident_text(expr: &Expr) -> Option<&str> {
+  match expr {
+    Expr::Ident(ident) => Some(&ident.sym),
+    _ => None,
+  }
+}
+
+fn get_member_prop_ident_text(member_prop: &MemberProp) -> Option<&str> {
+  match member_prop {
+    MemberProp::Ident(ident) => Some(&ident.sym),
+    _ => None,
+  }
+}
+
+fn get_member_prop_text(member_prop: &MemberProp) -> Option<&str> {
+  match member_prop {
+    MemberProp::Ident(ident) => Some(&ident.sym),
+    MemberProp::Computed(computed) => match &*computed.expr {
+      Expr::Lit(Lit::Str(str)) => Some(&str.value),
+      _ => None,
+    },
+    _ => None,
+  }
+}
+
+fn get_prop_name(prop: &Prop) -> Option<&str> {
+  match prop {
+    Prop::KeyValue(prop) => prop_name_from_key(&prop.key),
+    Prop::Shorthand(ident) => Some(&*ident.sym),
+    Prop::Getter(prop) => prop_name_from_key(&prop.key),
+    Prop::Setter(prop) => prop_name_from_key(&prop.key),
+    Prop::Method(prop) => prop_name_from_key(&prop.key),
+    // invalid for object literal, so ignore
+    Prop::Assign(_) => None,
+  }
+}
+
+fn prop_name_from_key(prop: &PropName) -> Option<&str> {
+  match prop {
+    PropName::Ident(ident) => Some(&*ident.sym),
+    PropName::Str(str) => Some(&*str.value),
+    PropName::BigInt(_) | PropName::Computed(_) | PropName::Num(_) => None,
+  }
+}
+
+#[cfg(test)]
+mod test {
+  use pretty_assertions::assert_eq;
+  use std::cell::RefCell;
+
+  use crate::parse_script;
+  use crate::MediaType;
+  use crate::ParseParams;
+  use crate::SourceTextInfo;
+
+  use super::*;
+
+  struct CjsAnalysisTester {
+    analysis: RefCell<CjsAnalysis>,
+  }
+
+  impl CjsAnalysisTester {
+    pub fn assert_exports(&self, mut values: Vec<&str>) {
+      values.sort_unstable();
+      let mut analysis = self.analysis.borrow_mut();
+      assert_eq!(analysis.exports, values);
+      analysis.exports.clear();
+    }
+
+    pub fn assert_reexports(&self, mut values: Vec<&str>) {
+      values.sort_unstable();
+      let mut analysis = self.analysis.borrow_mut();
+      assert_eq!(analysis.reexports, values);
+      analysis.reexports.clear();
+    }
+
+    pub fn assert_empty(&self) {
+      let analysis = self.analysis.borrow();
+      if !analysis.exports.is_empty() {
+        panic!("Had exports: {}", analysis.exports.join(", "))
+      }
+      if !analysis.reexports.is_empty() {
+        panic!("Had reexports: {}", analysis.reexports.join(", "))
+      }
+    }
+  }
+
+  impl Drop for CjsAnalysisTester {
+    fn drop(&mut self) {
+      // ensures that all values have been asserted for
+      if !std::thread::panicking() {
+        self.assert_empty();
+      }
+    }
+  }
+
+  fn parse_cjs(source: &str) -> CjsAnalysisTester {
+    let parsed_source = parse_script(ParseParams {
+      specifier: "".to_string(),
+      source: SourceTextInfo::from_string(source.to_string()),
+      media_type: MediaType::Cjs,
+      capture_tokens: true,
+      scope_analysis: false,
```

Comment:
This doesn't use scope analysis to tell if identifiers like `exports` has some declaration like `const exports = ...` in its scope. It would be trivial for us to add, but cjs-module-lexer doesn't do it and it requires an extra pass on the file to do scope analyis so I'm not sure it's worth the effort (at least for the first pass).